### DISCUSSION TITLE
Adding an "eject" command

### DISCUSSION THREAD

**Discussion by federicobond** [aws/copilot-cli#5657] [Ideas]
Let me start by saying that Copilot is wonderful. I've been following its development for a while and currently using it in prod. A tool like this has been missing in the AWS ecosystem for years. Copilot provides the fastest way to quickly set up a good system with modern, well-thought-out defaults.

It's however quite clear to me that once a project infrastructure achieves a certain level of complexity, Copilot will start to become a bit of a straightjacket, and it's defaults might not suit everyone. It would be awesome to provide a mechanism to then "eject" the project to CDK apps, and continue building it with IaC which gives way more flexibility during more advanced stages of development.

**Comment by iamhopaul123** [comment]
Hello @federicobond. We've been actually seriously thinking about this idea. In what abstraction level of an CDK app would you expect Copilot to be ejected to? Would you expect to see an app with all L2s? Or would you expect L3s that expose everything and allow users to extend?

**Reply by federicobond** [reply]
Great question. I'm honestly not sure what's best. I'd probably want to organize them in L3 constructs to avoid duplication if I have several applications for example, but those constructs can be emitted during the eject too, instead of imported from a lib. This would make iterating the eject output faster.

On the other hand, having those L3 imported from a lib means that customization points developed for later versions of Copilot can still be used by users with ejected systems but it can be harder to maintain compatibility while evolving them along with Copilot.

**Reply by iamhopaul123** [reply]
Yeah I guess that's the major concern - ejecting basically means everything will stop getting maintained by Copilot. For L1/L2s I guess there's already a path right now which is simply don't use Copilot to make any deployment - instead, start using CDK by migrating from CFN https://aws.amazon.com/blogs/devops/announcing-cdk-migrate-a-single-command-to-migrate-to-the-aws-cdk/
